bpo-42673 prevent branch misprediction in round_size (used in rehash)

jneb opened the pull request:
Replace the loop in round_size using bit_length to prevent branch misprediction delays.

<!-- issue-number: [bpo-42673](https://bugs.python.org/issue42673) -->
https://bugs.python.org/issue42673
<!-- /issue-number -->

jneb committed:
Update hashtable.c

Replace the loop in round_size using bit_length to prevent branch misprediction delays.

blurb-it[bot] committed:
📜🤖 Added by blurb_it.

jneb committed:
Update hashtable.c

Forgot the #linclude...

jneb committed:
Update hashtable.c

Updated comment.
I hope the new build test doesn't fail. (That wasn't my fault!)

vstinner commented on Python/hashtable.c:
```diff
@@ -109,10 +110,9 @@ round_size(size_t s)
-    return i;
+    // 1 << _Py_bit_length(s) is the smallest value k so that 1 << k > s
+    // subtract one in case s is an exact power of two, saving space
+    return 1 << _Py_bit_length(s - 1);
```
Can you please try _Py_SIZE_ROUND_DOWN() which could be even faster? It uses simple operations:

```
#define _Py_SIZE_ROUND_DOWN(n, a) ((size_t)(n) & ~(size_t)((a) - 1))
```

About _Py_bit_length(): it has a special case for 0, but s >= HASHTABLE_MIN_SIZE.

jneb replied:
That's brilliant! I didn't know about that macro.
EDITED:
.... but, that macro doesn't do what we want: we want to round up to the nearest power of two, not to a multiple of a given power of two.
I've been wrecking my brain for a quick binary trick to do that, but this is the fastest one I could find.

vstinner replied:
For me, this macro is pure black magic... but oops, I picked the wrong macro, I was thinking at: ``_Py_SIZE_ROUND_UP()``.

```
/* Below "a" is a power of 2. */
/* Round down size "n" to be a multiple of "a". */
#define _Py_SIZE_ROUND_DOWN(n, a) ((size_t)(n) & ~(size_t)((a) - 1))

/* Round up size "n" to be a multiple of "a". */
#define _Py_SIZE_ROUND_UP(n, a) (((size_t)(n) + \
        (size_t)((a) - 1)) & ~(size_t)((a) - 1))
```

vstinner replied:
Maybe _Py_SIZE_ROUND_UP() should be rewritten as a static inline function to ensure that (a-1) expression is only computed once.

vstinner commented on Python/hashtable.c:
```diff
@@ -109,10 +110,9 @@ round_size(size_t s)
     size_t i;
     if (s < HASHTABLE_MIN_SIZE)
```
Micro-optimisation :-)

```suggestion
    if (s <= HASHTABLE_MIN_SIZE)
```

jneb replied:
This actually may make the code slower. The value passed to round_size will almost never be < HASHTABLE_MIN_SIZE, but it sometimes is equal.
The branch prediction will probably be faster if it can assume the condition is always false.

jneb committed:
Removed unused variable

In the last patch, I forgot to remove an unused variable.

jneb reviewed (commented):
That's a good point. So this would fail if the hashtable gets a number of entries that couldn't fit in an unsigned long, which is indeed possible in theory (but only for humongous computers: 8<<32 = 32 GB, for a single hash table. Would be a bit hard to test :-).
I think the best solution would be to fix the macro to handle this case.

jneb committed:
Added a separate bit_length for size_t

This is needed for the new version of the hashtable.c, in case the hashtable is extremely large (>32GB).

jneb committed:
Fixed sizeof in #if

Apperently, you can't use sizeof in an #if

jneb committed:
Another attempt to fix

There should be a good way to check sizeof(size_t) > sizeof(unsigned long); I hope this is the one

jneb committed:
Silly typo fixed

Swapped a } and an #endif...

jneb committed:
Separate routine _Py_bit_length_size_t

Made a separate routine _Py_bit_length_size_t since overloading doesn't seem to work

jneb committed:
Separate _Py_bit_length_size_t

Use the separate _Py_bit_length_size_t so that round_size works in extreme cases of hashtables >32GB

jneb committed:
Simple fix

Cast the 1 to size_t to get a proper shift

jneb committed:
Fixed type of msb

I changed one size_t too many

jneb committed:
Use 64 bit bsr

I am determined to get this to work!

vstinner commented on Include/internal/pycore_bitutils.h:
```diff
@@ -170,6 +170,49 @@ _Py_bit_length(unsigned long x)
+// Same routine as above, for size_t, if that is bigger than unsigned long
+// This means bitscanreverse (as used above) is not going to work
+static inline int
+_Py_bit_length_size_t(size_t x)
```
I would prefer to add a _Py_next_pow2() function which takes a size_t parameter. So we can use more efficient implementation when __builtin_clzl() and _BitScanReverse() are not available. Like:

```
static inline uint32_t next_pow2(uint32_t x)
{
	x -= 1;
	x |= (x >> 1);
	x |= (x >> 2);
	x |= (x >> 4);
	x |= (x >> 8);
	x |= (x >> 16);
	return x + 1;
}
```

I suggest to require the argument to be ``>= 2``.

vstinner replied:
This code is reference at http://graphics.stanford.edu/~seander/bithacks.html#RoundUpPowerOf2 for example.
